homeassistant/components/freedompro/climate.py

```diff
@@ -0,0 +1,144 @@
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Freedompro climate."""
+    api_key = entry.data[CONF_API_KEY]
+    coordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        Device(hass, api_key, device, coordinator)
+        for device in coordinator.data
+        if device["type"] == "thermostat"
+    )
+
+
+class Device(CoordinatorEntity, ClimateEntity):
+    """Representation of an Freedompro climate."""
+
+    def __init__(self, hass, api_key, device, coordinator):
+        """Initialize the Freedompro climate."""
+        super().__init__(coordinator)
+        self._session = aiohttp_client.async_get_clientsession(hass)
+        self._api_key = api_key
+        self._attr_name = device["name"]
+        self._attr_unique_id = device["uid"]
+        self._characteristics = device["characteristics"]
+        self._attr_device_info = {
+            "name": self.name,
+            "identifiers": {
+                (DOMAIN, self.unique_id),
+            },
+            "model": device["type"],
+            "manufacturer": "Freedompro",
+        }
+        self._attr_supported_features = SUPPORT_TARGET_TEMPERATURE
+        self._attr_current_temperature = 0
+        self._attr_target_temperature = 0
+        self._attr_temperature_unit = TEMP_CELSIUS
+        self._attr_hvac_mode = HVAC_MODE_OFF
+        self._attr_hvac_modes = SUPPORTED_HVAC_MODES
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        device = next(
+            (
+                device
+                for device in self.coordinator.data
+                if device["uid"] == self._attr_unique_id
+            ),
+            None,
+        )
+        if device is not None and "state" in device:
+            state = device["state"]
+            if "currentTemperature" in state:
+                self._attr_current_temperature = state["currentTemperature"]
+            if "targetTemperature" in state:
+                self._attr_target_temperature = state["targetTemperature"]
+            if "heatingCoolingState" in state:
+                self._attr_hvac_mode = HVAC_MAP[state["heatingCoolingState"]]
+        super()._handle_coordinator_update()
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+        self._handle_coordinator_update()
+
+    async def async_set_hvac_mode(self, hvac_mode):
+        """Async function to set mode to climate."""
+        if hvac_mode not in SUPPORTED_HVAC_MODES:
+            _LOGGER.error(
+                "Got unsupported hvac_mode %s, expected one of %s",
+                hvac_mode,
+                SUPPORTED_HVAC_MODES,
+            )
+            return
+
+        payload = {}
+        payload["heatingCoolingState"] = HVAC_INVERT_MAP[hvac_mode]
+        payload = json.dumps(payload)
```

Comment:
Dumping the payload looks like something that should be handled by the library. Please improve this in a follow up PR.

Generally it's nicer for library users to not have to deal with "payloads" at all but instead deal with class instances that represent a part of a library feature, eg a device, and instance attributes that represent sub features, and have methods that update the state of the instance. I recommend using a data validation library, eg pydantic, to help with this.

Reply:
ok thanks I will do it in a next PR